### ISSUE TITLE
[ENH]: API discussion for grouped bar charts

### ISSUE THREAD

**Issue by timhoffm** [matplotlib/matplotlib#24313]
### Problem

Currently, if one wants to draw multiple categories of bars side-by-side as in https://matplotlib.org/devdocs/gallery/lines_bars_and_markers/barchart.html, one has to calculate the bar positions manually. This is really a nuissance and too low level for any user-facing interface. Actually, when asked I recommend to use pandas plotting functions for that if possible, which is really embarrasing.

There have been stalled attempts to do this (issure 10610, PR #11048). Additionally, this often gets related to a function for stacked bars #14086.

### Proposed solution

I'd like to pick up this topic and come up with a reasonable API. This topic is really complex and one can easily get lost in various details. For the design procedure, I take a bottom-up appoach by starting with a basic function `grouped_bar()` that does only expose the minimal functionality to get the plot done. I then intended to additional parameters one by one as they fit in. - So don't be concerned that the first proposal here is quite basic.

**Terminology** I'll use *label* for the x-values, i.e. 'G1' .. 'G5' in above example, and *group* for the categories, i.e. 'Tea'/'Coffee'.

---

###  For now: Only vertical orientation
I'll limit the discussion here to 'vertical'. We can have a separate discussion whether we want to add an *orientation* parameter or make a `grouped_barh`. Both are technically easy and only an API design decision that's orthogonal to the rest of the API.

### For now: Only grouped layout, no stacked layout
To keep things simple, I'll limit myself to grouped for now, because:
- Stacked is somewhat simpler than grouped as you only need to insert the *bottom* values in multiple calls. For two bars it's just the heights of the firsts, and a cumsum for more than two.
- We may build a separate `stacked_bar()` function, if the first bullet point is considered too cumbersome.
- It's conceivable to unite both in one function as poposed in #11048 and realized in `DataFrame.plot.bar`

either way let's defer stacked bars to later.

### Minimal API

We want to be able to rewrite https://matplotlib.org/devdocs/gallery/lines_bars_and_markers/barchart.html as

```
grouped_bar(labels, [tea_means, coffee_means], group_labels=['Tea', 'Coffee']) 
```

Thus, the minimal API is:

```
def grouped_bar(x, heights, *, group_labels=None):
    """
    Parameters
    -----------
    x : array-like of str
         The labels.
    heights : list of array-like:
         An iterable of array-like: The iteration runs over the groups.
         Each individual array-like is the list of label values for that group.
    group_labels : array-like of str, optional
         The labels of the data groups.
    """
```

---

I'll soon expand on the minimal API. But before that, please speak up in case you have fundamental concerns with adding such functionality at all or with the bottom-up design approach. OTOH if you think this is worth pursuing, please give a :+1:.

**Comment by timhoffm** [comment]
@story645 I have answers (and opinions :smile:) on all of these. I will hold off diving into the details for a bit to see whether there is general high-level feedback.

**Comment by mwaskom** [comment]
No particular thoughts here; I doubt that seaborn would use this API.

**Comment by oscargus** [comment]
I think it would be nice (although I realize pretty hard) to have a single bar-call that can do grouped and stacked. Or at least be able to relatively easy do both grouped and stacked (as in the same bar graph). I cannot say that I have a particular use case in mind, but considering that plots tend to run out of dimensions it may be worth considering if we are going for it.

I have not read up, but I think the 2D-path is probably to be preferred? Now, of course that can be introduced via the `grouped_bar` approach. (For stacked and grouped, 3D?)

**Comment by tacaswell** [comment]
I do think we will want to handle stacked vs adjacent eventually (bokeh calls adjacent "doged" https://docs.bokeh.org/en/latest/docs/gallery/bar_dodged.html).  In both cases you are answering the question of "I have N bars that should be 'at the same point' how do I de-conflict them?".  Another reason to merge the two cases into one is so that they can have a more coherent story around how to make legend entries for the set at once.  Should probably do this with an eye for how to both stack and dodge?

I am 👍🏻 on this (as I still agree with what I wrote in the linked 2018 comment even though I have no memory of writing it) over all.

**Comment by timhoffm** [comment]
All agreed. It should be fairly to have a `multi_bar()` method that that can handle grouped or stacked bars. I'm more sceptical on grouped+stacked because that adds an additional dimension to the inputs. - Likely the best solution for that is still a sequence of grouped bar plots with offsets (similar what one has currently do for stacking).

**Comment by jklymak** [comment]
Where our api gets in trouble is guessing what the user wants.  I'd have two methods for these two concepts.

**Comment by mwaskom** [comment]
Is it worth mentioning the precedent that `hist` has these as options within the same method?

**Comment by mwaskom** [comment]
Heh. Yeah, for consistency this would suggest there should be only one method, but then again I think the `hist` API is kind of a cautionary tale...

**Comment by mwaskom** [comment]
> we have .bar and .barh and that could easily be one method w/ an orientation parameter...

Ah, the way `hist` does it! :D

**Comment by timhoffm** [comment]
If viable, parameters have the advantage of not leading to combinatory explosion.

We likely don't want end up with

- `bar()`
- `barh()`
- `grouped_bar()`
- `grouped_barh()`
- `stacked_bar()`
- `stacked_barh()`

**Comment by jklymak** [comment]
Bar/barh seem unnecessary to me since it is just changing the major axis.  But grouping or stacking seem unique enough with their own set of options that they should get their own methods.

**Comment by tacaswell** [comment]
As an implementation detail, `barh` is in fact a thin wrapper of `bar` making use of an (undocumented) kwarg to control the orientation.  Have not gone back to look at the discussion, but my guess is we did that to reduce duplicate code, maintain back compatibility, and not provide multiple (documented) ways to do the same thing.  I think we should go with Tim's initial comments and not get hung up on horizontal vs vertical at this point and table that whole discussion until we have vertical working (but maybe keep in mind names that are not deeply directional (like "base" instead of "bottom") if we are willing to not directly track `bar`).

I agree `hist` is a cautionary tale of doing too much in a single function.

**Comment by mwaskom** [comment]
`barh` is probably a Matlab compatibility thing

**Comment by mwaskom** [comment]
BTW, not exactly an API issue but something you're going to run into as you implement this that I'll flag since it's been a headache for me.

Because half of the edges lie outside of the data limits of a patch artist, dodged/stacked bars can look strange when they have different edge colors. e.g.:

```python
f, ax = plt.subplots()
x = np.array([0, 1])
ax.bar(x - .2, [1, 2], .4, linewidth=3, facecolor=".8", edgecolor=".2")
ax.bar(x + .2, [1.5, 1], .4, linewidth=3, facecolor="none", edgecolor="C1")
```
![image](https://user-images.githubusercontent.com/315810/199703162-0ad59ee7-3f77-48e4-a1a0-015a337d1a24.png)

Note that, rather than abutting, the orange edges fully cover the black edges.

**Comment by timhoffm** [comment]
Did you try to solve this, or do you even have a solution ready? I think this is not trivial as the edge and data have different units, so even fine-tuning the distance would be brittle. IMHO the only proper solution would be a "draw line on the inside" concept, which OTOH results in fat lines if the edges have the same color.

I think for now, we have to live with the overlap.

**Comment by mwaskom** [comment]
Agreed, this can’t be handled easily by positional adjustments. I think it needs to be solved in the way the path artist is drawn, but for now my workaround is to set a clip path on the patch. This works but has some weird artifacts in raster plots (AFK or else I’d demonstrate).

**Comment by tacaswell** [comment]
_very_ offtopic:

This is the issue with path stroking issue that came up breifly when we were talking about the AWS ask.

The issue here is that when we stroke a path the notional infinitly thin path goes down the center of the drawn line, with half of the line width on either side.  What I think we want here is the ability to stroke the path either inside our outside.  On very quick search it looks like this was proposed in for SVG but not implemented (https://www.w3.org/TR/SVG/painting.html#SpecifyingStrokePaint).  We could probably do this at draw time if we have a algorithm to shrink or grow an arbitrary path perpendicular to itself in screen coordinates (which we have very late in the draw process).

**Comment by mwaskom** [comment]
Ah, yeah, if this is the behavior of SVG primitives then I guess there's not much to do about it.

**Comment by tacaswell** [comment]
I strongly suspect the issue is that in the second case the lines are an odd-handful of pixel wide, the lines pixel-snap in Agg (to avoid anti-aliasing thin lines and making them look "fuzzy") and I would assume that other renderers do as well.  If the clip path is similarly snapped then one side of the rectangles can be 1 pixel narrower and with very thin lines that is enough to catch our eye.  The same effect is probably happening with the thicker lines but we are just less sensitive to it.

For the examples I'm guessing "do not draw the edges" is not an option?

**Comment by QuLogic** [comment]
Is there any more to be done here now that #28560 is in?

**Comment by timhoffm** [comment]
No, this can be closed.
